File: atcoder_selected_100/kotlin/dpl_1_i.kt
/**
 * -
 * DPL-1/I
 * 動的計画法:ナップサックDP
 * 個数制限付きナップサックDP
 * 貪欲法
 * スライド最小値
 */
 fun main() {
  val (N, W) = readLine()!!.split(" ").map(String::toInt)
  val VWM = List(N) { readLine()!!.split(" ").map(String::toInt) }
  val maxV = VWM.maxOf { it[0] }
  val totalV = VWM.sumOf { (v, _, m) -> v * minOf(m, maxV) }

  val dp = IntArray(totalV + 1) { W + 1 }
  dp[0] = 0

  VWM.forEach { (v, w, m) ->
    val limit = minOf(m, maxV)
    repeat(v) { remainder ->
      val deque = ArrayDeque<Pair<Int, Int>>()
      var j = 0
      while (j * v + remainder <= totalV && j * w <= W) {
        val value = j * v + remainder
        val reducedW = dp[value] - j * w
        while (deque.isNotEmpty() && deque.last().second > reducedW) deque.removeLast()
        while (deque.isNotEmpty() && deque.first().first < j - limit) deque.removeFirst()
        deque.addLast(j to reducedW)
        dp[value] = minOf(dp[value], deque.first().second + j * w)
        j++
      }
    }
  }

  val remainingVwm = VWM.map { (v, w, m) -> Triple(v, w, maxOf(m - maxV, 0)) }
    .filter { it.third > 0 }
    .sortedBy { (v, w, _) -> w.toDouble() / v }

  var max_value = 0
  dp.forEachIndexed { value, weight ->
    if (weight > W) return@forEachIndexed
    var total_value = value
    var remainingW = W - weight
    remainingVwm.forEach { (v, w, m) ->
      val count = minOf(m, remainingW / w)
      total_value += count * v
      remainingW -= count * w
    }
    max_value = maxOf(max_value, total_value)
  }
  println(max_value)
}

File: atcoder_selected_100/ruby/dpl_1_i.rb
# -
# 動的計画法:ナップサックDP
# 個数制限付きナップサックDP
# 貪欲法
# スライド最小値

N, W = gets.split.map(&:to_i)
VWM = Array.new(N) { gets.split.map(&:to_i) }

max_v = VWM.max_by(&:first)[0]
total_v = VWM.sum { |v, _, m| v * [max_v, m].min }

# dp[i]: 価値が i のときの重さの最小値
dp = Array.new(total_v + 1, W + 1)
dp[0] = 0

VWM.each do |v, w, m|
  limit = [max_v, m].min
  # スライド最小値を用いて、価値が i のときの重さの最小値を求める
  v.times do |other_v|
    deque = []
    j = 0
    while (value = j * v + other_v) <= total_v && j * w <= dp[value]
      other_w = dp[value] - j * w
      deque.pop while !deque.empty? && deque[-1][1] > other_w
      deque.shift while !deque.empty? && limit < j - deque[0][0]
      deque << [j, other_w]
      dp[value] = [dp[value], deque[0][1] + j * w].min
      j += 1
    end
  end
end

# 残りの容量に対して、価値あたり重さの小さい順に貪欲法で詰める
remaining_vwm = VWM.map {|v, w, m| [v, w, m - max_v] if m > max_v }.compact.sort_by {|v, w, _| w.to_f / v }

max_value = 0
dp.each_with_index do |weight, i|
  next if weight > W
  total_value = i
  remaining_w = W - weight
  remaining_vwm.each do |v, w, m|
    take = [m, (remaining_w / w).to_i].min
    total_value += v * take
    remaining_w -= w * take
  end
  max_value = [max_value, total_value].max
end

puts max_value
